go/base_client.go

```diff
@@ -387,6 +388,173 @@ func toCStrings(args []string) ([]C.uintptr_t, []C.ulong) {
 	return cStrings, stringLengths
 }
 
+func (client *baseClient) executeBatch(
+	ctx context.Context,
+	batch pipeline.Batch,
+	raiseOnError bool,
+	options *pipeline.BatchOptions,
+) ([]any, error) {
+	// Check if context is already done
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		// Continue with execution
+	}
+	// make the channel buffered, so that we don't need to acquire the client.mu in the successCallback and failureCallback.
+	resultChannel := make(chan payload, 1)
+	resultChannelPtr := unsafe.Pointer(&resultChannel)
+
+	pinner := pinner{}
+	pinnedChannelPtr := uintptr(pinner.Pin(resultChannelPtr))
+	defer pinner.Unpin()
+
+	client.mu.Lock()
+	if client.coreClient == nil {
+		client.mu.Unlock()
+		return nil, &errors.ClosingError{Msg: "ExecuteCommand failed. The client is closed."}
+	}
+	client.pending[resultChannelPtr] = struct{}{}
+
+	batchInfo := createBatchInfo(pinner, batch)
+	var optionsPtr *C.BatchOptionsInfo
+	if options != nil {
+		batchOptionsInfo := createBatchOptionsInfo(pinner, *options)
+		optionsPtr = &batchOptionsInfo
+	}
+
+	C.batch(
+		client.coreClient,
+		C.uintptr_t(pinnedChannelPtr),
+		&batchInfo,
+		C._Bool(raiseOnError),
+		optionsPtr,
+	)
+	client.mu.Unlock()
+
+	// Wait for result or context cancellation
+	var payload payload
+	select {
+	case <-ctx.Done():
+		client.mu.Lock()
+		if client.pending != nil {
+			delete(client.pending, resultChannelPtr)
+		}
+		client.mu.Unlock()
+		return nil, ctx.Err()
+	case payload = <-resultChannel:
+		// Continue with normal processing
+	}
+
+	client.mu.Lock()
+	if client.pending != nil {
+		delete(client.pending, resultChannelPtr)
+	}
+	client.mu.Unlock()
```

Comment:
The same pattern now occurs multiple times in the code (the only difference being the value that is returned upon context cancellation).
Perhaps this should be refactored into 
```
func(client *baseClient) waitForPayload(ctx, resultChannel, resultChannelPtr) (payload, error) {
	defer func() {
		client.mu.Lock()
		if client.pending != nil {
			delete(client.pending, resultChannelPtr)
		}
		client.mu.Unlock()
	}()
	
	// Wait for result or context cancellation
	var payload payload
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case payload = <-resultChannel:
		// Continue with normal processing
	}
	return payload, nil
}
```

Reply:
Though on second thought the cleanup of `resultChannelPtr` should be kept together with the logic that creates the channel and pins it, not be factored out into an unrelated method.